Implement filters for Dag run

RoyLee1224 opened the pull request:
## Related issues

closes: #53043


## To-Do: 
**Test coverage:**
- [x] Add test to `test_dag_run.py` to verify new `conf_contains` & `duration` filters

**Filters:**
- [x] Dag ID
- [x] Run ID
- [x] Start Date
- [x] End Date
- [x] Run After (Range)
- [x] Triggered By
- [x] Dag Version and/or Bundle Version
- [x] Duration Range (currently using from & to, improved `range selection` component would be a separate PR)
- [x] Conf Content
- [x] Delete `triggeringUserNameFilter` translation keys
- [x] Move all filters into `filterBar`
- [ ] ~~Producing Asset~~

## Screenshots
<img width="1095" height="624" alt="CleanShot 2025-09-16 at 16 58 37" src="https://github.com/user-attachments/assets/d5702592-2e95-4f3a-89e9-31c0c17ea46e" />


https://github.com/user-attachments/assets/5d71a2ca-04a3-4973-a872-b4c7919b74a4

## Follow up PR
- Add new type filters: `number range` , `date range`, `tag`

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

RoyLee1224 committed:
feat(ui): implement filterbar for Dag run

feat: add filter for dag Version

feat(api): Add duration filter for Dag runs

feat(ui): implement duration filter for Dag runs

feat(ui,api): implement conf_contains filter

refactor(ui): move trigering user filter into filterbar

refactor: Simplify JSON type detection for filtering

fix(i18n): modify triggering user translation

jason810496 commented on airflow-core/src/airflow/api_fastapi/common/parameters.py:
```diff
@@ -363,6 +364,13 @@ def to_orm(self, select: Select) -> Select:
+            # For JSON/JSONB columns, convert to text before applying LIKE
+            from sqlalchemy import JSON, Text, cast
+
+            if isinstance(self.attribute.type, JSON) or str(self.attribute.type).upper() in ("JSON", "JSONB"):
```
Just to make double check: Do we still need `str(self.attribute.type).upper() in ("JSON", "JSONB")` because `isinstance(self.attribute.type, JSON)` doesn't work for Postgres?

Or maybe `isinstance(self.attribute.type, JSON)` can handle all database dialects?

RoyLee1224 replied:
I think we do need `str(self.attribute.type).upper() in ("JSON", "JSONB")`, the query fails if we only use `isinstance(self.attribute.type, JSON)`

RoyLee1224 replied:
The error:

```
sqlalchemy.exc.ProgrammingError: (psycopg2.errors.UndefinedFunction) operator does not exist: jsonb ~~ text
HINT: You might need to add explicit type casts.
```

jason810496 replied:
Thanks for the explainaton! If this is the case, it seems `str(self.attribute.type).upper() in ("JSON", "JSONB")` will be enough?

RoyLee1224 replied:
Oh that makes sense! I'll fix it right away.

jason810496 reviewed (commented):
Nice! Thanks for the PR! 
LGTM on the API side and it would be nice to add the test in `airflow-core/tests/unit/api_fastapi/core_api/routes/public/test_dag_run.py` to verify the new filters, thanks!

bbovenzi commented on airflow-core/src/airflow/ui/src/constants/filterConfigs.tsx:
```diff
@@ -128,6 +153,12 @@ export const useFilterConfigs = () => {
+    [SearchParamsKeys.TRIGGERING_USER_NAME_PATTERN]: {
+      hotkeyDisabled: true,
+      icon: <FiUser />,
+      label: translate("common:filters.triggeringUser", "Triggering User"), // To do: add translation key and remove default value
```
We already have `common:dagRun.triggeringUser`

RoyLee1224 replied:
Perhaps `Triggering User Name` is a little bit too long in dropdown?

bbovenzi replied:
It is. I think we should remove "Name" from that existing translation

RoyLee1224 replied:
Sure, working on it

RoyLee1224 committed:
fix: resolve conflicts & move all filters into filterBar

RoyLee1224 committed:
test: Add test for duration and conf_contains filters

test: modify duration test

RoyLee1224 committed:
fix(ui): Add icons and styled options to Dag runs select type filters

fix(tests): Correct  assignment in  fixture

fix: Correct tests for duration and conf_contains filters

fix(test): using from_datetime_to_zulu_without_ms for end date

fix(tests): Correct tuple assignment for  in test setup

RoyLee1224 committed:
fix(tests): Correct datetime formatting and test expectations in dag_run API tests

RoyLee1224 committed:
fix(tests): Simplify conf_contains filter test patterns

pierrejeambrun commented on airflow-core/tests/unit/api_fastapi/core_api/routes/public/test_dag_run.py:
```diff
@@ -178,23 +194,45 @@ def get_dag_versions_dict(dag_versions: list[DagVersion]) -> list[dict]:
+
+    This function ensures test expectations match the actual API behavior.
+    If API datetime serialization changes in the future, only this function needs updating.
+
```
I never realized that.

IT's super weird.

RoyLee1224 replied:
To be honest, I'm getting some odd test results caused by the time formats.

Still trying to figure it out. I'll update the test after having a better understand with it.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/constants/filterConfigs.tsx:
```diff
@@ -61,6 +79,31 @@ export const useFilterConfigs = () => {
+    },
+    [SearchParamsKeys.DURATION_GTE]: {
+      icon: <MdHourglassEmpty />,
+      label: translate("common:filters.durationFrom", "Duration From"), // To do: add translation key and remove default value
```
Did we miss these two 'todos' ?

The PR is marked for `3.2.0`, this will not break translation freeze. You can go ahead and do that change.

RoyLee1224 replied:
Sure, I'm on it.

RoyLee1224 replied:
Actually, I'd like to propose we hold off on adding specific translation keys for "Duration From" and "Duration To" for now.

The reason is that I'm developing `range-selection` component, which will only use "From" and "To." Adding the keys now would create temporary work for language owners, as those keys would need to be removed once the new component is merged.

RoyLee1224 replied:
Something like this:

<img width="3024" height="1798" alt="CleanShot 2025-09-24 at 13 48 16@2x" src="https://github.com/user-attachments/assets/c8954dac-7e9d-443e-9b4e-4dd7cbfb8de3" />

pierrejeambrun replied:
> Adding the keys now would create temporary work for language owners, as those keys would need to be removed once the new component is merged.

To be realistic I don't think any translation owner will work on translation for `3.2.0` now. So even if we merge this it's very unlikely that people will start translation for that.

If we do it the other way around though, the risk is that we forget to do the translation for them, and that is worse I think.


I'm in favor of adding "Duration From" and "Duration To" translation keys now, and you can remove that in your follow up PR for `date-selection` component. (Also maybe other keys will be removed at that time)

RoyLee1224 replied:
Sure, working on it!

pierrejeambrun reviewed (approved):
Looking good to me.  

Just one question.

jason810496 reviewed (commented):
Thanks for the update!
Same comment as Pierre, but the CI still need fix, thanks!

RoyLee1224 committed:
test: using int for testing duration

RoyLee1224 committed:
Merge branch 'main' into dag_run_filter

RoyLee1224 committed:
fix(models): Make DagRun.duration property SQLite-compatible

RoyLee1224 committed:
Merge branch 'main' into dag_run_filter

jason810496 reviewed (approved):
>  Fixed the DagRun.duration hybrid property to support the SQLite dialect, which was causing the test failures. It now uses julianday() for calculations. CI should pass now!

Nice catch! Thanks for the fix.
LGTM on API side.

RoyLee1224 committed:
fix(i18n): add duration translation keys

RoyLee1224 committed:
Merge branch 'main' into dag_run_filter

pierrejeambrun reviewed (approved):
LGTM thanks
